diff --git a/src/Plugins/Qt/QTMWidget.cpp b/src/Plugins/Qt/QTMWidget.cpp
index a7f88df4f..7fdaa1f96 100644
--- a/src/Plugins/Qt/QTMWidget.cpp
+++ b/src/Plugins/Qt/QTMWidget.cpp
@@ -498,12 +498,12 @@ QTMWidget::inputMethodEvent (QInputMethodEvent* event) {
   
   if (!commit_string.isEmpty()) {
     bool done= false;
+    string s= from_qstring (commit_string);
 #ifdef OS_MACOS
 #if (QT_VERSION < 0x050000)
     // NOTE: this hack is only needed for Qt4 under MacOS,
     // but it only works for standard US keyboards
     done= true;
-    string s= from_qstring (commit_string);
     Qt::KeyboardModifiers SA= Qt::ShiftModifier | Qt::AltModifier;
     if (s == "\17") kbdEvent (36, Qt::AltModifier, commit_string);
     else if (s == "<ddagger>") kbdEvent (38, Qt::AltModifier, commit_string);
@@ -524,6 +524,12 @@ QTMWidget::inputMethodEvent (QInputMethodEvent* event) {
 #endif
 #endif
     
+#ifdef OS_MINGW
+    if (commit_string.size() == 1 && is_alpha(s[0])) {
+      done= true; 
+    }
+#endif
+
     if (!done) {
       if (DEBUG_QT)
         debug_qt << "IM committing: " << commit_string.toUtf8().data() << LF;
